src/ProtocolGateway.Core/Mqtt/MqttAdapter.cs

```diff
@@ -1107,7 +1107,8 @@ bool IsReadAllowed()
                 var ex = task.Exception.InnerException as ChannelMessageProcessingException;
                 if (ex != null)
                 {
-                    ShutdownOnError(ex.Context, scope, task.Exception);
+                    // we need to unmask the ChannelMessageProcessingException to get the real exception
+                    ShutdownOnError(ex.Context, scope, ex.InnerException);
```

Comment:
why is it necessary here? by doing so we'd be erasing the exception in the chain and it also subtly introduces a convention that ChannelMessageProcessingException must have an inner exception. Shutdown can be triggered due to multiple reasons so having an idea through exception type may be valuable for downstream handlers.
Can you instead unpack the exception where you have enough context to do so?

Reply:
It should always have an inner exception, which is masked by ChannelMessageProcessingException in https://github.com/Azure/azure-iot-protocol-gateway/blob/master/src/ProtocolGateway.Core/Mqtt/MessageAsyncProcessorBase.cs#L109

And given line 1107&1108 shutdown only happens when there is a ChannelMessageProcessingException with an exception it masks. The problem we are facing is that the mask by ChannelMessageProcessingException makes caller (Gateway client) hard to track the real exception. Caller only cares about the exception today which is a generic ChannelMessageProcessingException so we can even throw server error even if inner exception is a client error.